### ISSUE TITLE
`rate` produces incorrect negative values

### ISSUE THREAD

**Issue by uhthomas** [VictoriaMetrics/VictoriaMetrics#5571]
### Describe the bug

The query `sum(rate(container_cpu_usage_seconds_total{image!=""}[$__rate_interval])) by (pod)` can produce the invalid value `-9.47e-16`.

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/9749173/b0df7972-8200-4539-a42e-259b6ef0a28a)

The raw values are never negative.

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/9749173/2c56378d-7607-45a0-ac06-43e4f97daba1)

### To Reproduce

Not sure - use the above query on a pod which is restarting a lot?

It seems to be `rate` which is an issue.

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/9749173/9818a593-e7ef-4f9b-ac89-d4161145f1f7)

### Version

v1.95.1

**Comment by Haleygo** [comment]
Hello!
`rate` should be able to handle counter resets, could be a corner case here.
What's your `$__rate_interval` here? And have you set `-dedup.minScrapeInterval`?
![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/39937150/13eed58b-550a-49b4-b986-507a262ddc7d)

And could you please [export](https://new.docs.victoriametrics.com/url-examples/index.html#apiv1export) raw data of this series, as I can't reproduce it with my data.

**Comment by uhthomas** [comment]
Rate interval is `1m0s`. A rate interval of `1m` does not produce this?

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/9749173/581867da-559c-4eea-8aba-00840c8565c4)

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/9749173/e4bc2ca9-9a5c-4a76-a32f-03a81cb102eb)

Please see attached the exported time series.

```sh
❯ curl http://localhost:8429/api/v1/export -d 'match[]=container_cpu_usage_seconds_total{pod="bazarr-0"}' > container_cpu_usage_seconds_total.json
```

https://github.com/uhthomas/vm5571/blob/main/container_cpu_usage_seconds_total.json

**Comment by Haleygo** [comment]
Thanks for your data!
I can reproduce the problem with them and I think it's related to float64 caculation precision problem when trying `removeCounterResets` for `rate()`
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/3a3b90f16a0609b2a4a534a7534852a35c5d7852/app/vmselect/promql/rollup.go#L841-L863

Like the result of `0.062454 + 41.57826100000001` will be `41.640715000000014` instead of `41.64071500000001`,  and it would be a problem in some rare cases, see a wrong test case in https://github.com/VictoriaMetrics/VictoriaMetrics/blob/c4b5d85c716e5cd279dd6eb2e8ef6169553c78dc/app/vmselect/promql/rollup_test.go#L162-L174

**Comment by Haleygo** [comment]
Thanks for your data!
I can reproduce the problem with them and I think it's related to float64 caculation precision problem when trying `removeCounterResets` for `rate()`
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/3a3b90f16a0609b2a4a534a7534852a35c5d7852/app/vmselect/promql/rollup.go#L841-L863

Like the result of `0.062454 + 41.57826100000001` will be `41.640715000000014` instead of `41.64071500000001`,  and it would be a problem in some rare cases[frequent reset and small window], see a wrong test case in https://github.com/VictoriaMetrics/VictoriaMetrics/blob/3a3b90f16a0609b2a4a534a7534852a35c5d7852/app/vmselect/promql/rollup_test.go#L162-L174

**Comment by uhthomas** [comment]
Cool, thanks for investigating. Could you also help me understand how `1m` and `1m0s` produce different results for the same query? That doesn't make sense to me.

**Comment by Haleygo** [comment]
> Cool, thanks for investigating. Could you also help me understand how `1m` and `1m0s` produce different results for the same query? That doesn't make sense to me.

It might related to result cache. If you have executed query using `1m` before with bigger timerange, the following query request with smaller timerange will use cached results, but change expr with `1m0s` will lead to cache missed.
To prove this, could you [disable cache](https://new.docs.victoriametrics.com/troubleshooting/?highlight=nocache%3D1#unexpected-query-results) and see if there still are different results?

**Comment by valyala** [comment]
VictoriaMetrics should stop returning negative values from `rate()` and `increase()` functions starting from [v1.97.0](https://docs.victoriametrics.com/CHANGELOG.html). Closing the issue as fixed then.

**Comment by valyala** [comment]
FYI, the bugfix for this issue has been included in [v1.87.14 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.87.14).
